### ISSUE TITLE
info-hash regardless of it being invalid

### ISSUE THREAD

**Issue by cfpp2p** [drizt/torrent-file-editor#47]
Editor sorts **_info_** dictionary upon save. Without any changes at all to **_original_** info dictionary it is always sorted upon save. Editing and making changes outside of info should not change original hash.

http://bittorrent.org/beps/bep_0003.html
> ...20-Oct-2012 (arvid@bittorrent.com), clarified that info-hash is the digest of en bencoding found in .torrent file...

> ## trackers
> Tracker GET requests have the following keys:
> info_hash
>     The 20 byte sha1 hash of the bencoded form of the info value from the metainfo file. Note that this is a substring of the metainfo file. The info-hash must be the hash of the encoded form as found in the .torrent file, **_regardless of it being invalid_**...

https://github.com/transmission/transmission/issues/129#issuecomment-270539435
> It's true that bencoded dictionary keys MUST be sorted.
> However, when some 'private trackers' started to generate erroneous .torrents, the chosen resolution was to hash the unsorted info dict as it is. BEP3 has been updated to reflect this.

**Comment by drizt** [comment]
Please show the problem. How to reproduce?

**Comment by cfpp2p** [comment]
[info-mis-ordered-before-save-torrent](http://transmissionbt.net/info-mis-ordered-before-save.zip)

Hashes via uTorrent, BitComet and (fixed) Transmission clients.
mis-ordered-before-save e00594d7edfd457708c10bc60cf63096a2b5d471
mis-ordered-after-save 3b27f588611cccca4fccacb860534b05604a62ea

Open the sample torrent with the editor, then immediately and simply save it. The info dictionary is reordered upon the editor opening the torrent file, thereby the altered hash when saving. If nothing is at all changed within info dictionary the hash shouldn't change according to the specs. Changes or additions outside the info dictionary never should make a difference to the hash if what is in info is not altered in any way.

The **"source"** string is not in order within info dictionary.

mis-ordered-before-save


d8:announce45:http://yourprivayetracker.com/?pass=1234567897:comment45:Please seed this torrent as long as possible!10:created by39:This torrent has been created by cfpp2p13:creation datei1475598235e4:infod6:source29:http://yourprivayetracker.com5:filesld6:lengthi3785877485e4:pathl14:Free State.mkveed6:lengthi95534e4:pathl4:Subs19:Spanish - Latin.srteed6:lengthi99638e4:pathl4:Subs10:French.srteed6:lengthi99786e4:pathl4:Subs11:English.srteed6:lengthi113036e4:pathl4:Subs16:English(SDH).srteee4:name10:Free State12:piece lengthi262144e6:pieces288880: ......
.....
:privatei1eee

.
.
.
mis-ordered-after-save

d8:announce45:http://yourprivayetracker.com/?pass=12345678913:announce-listll45:http://yourprivayetracker.com/?pass=123456789ee7:comment45:Please seed this torrent as long as possible!10:created by39:This torrent has been created by cfpp2p13:creation datei1475598235e4:infod5:filesld6:lengthi3785877485e4:pathl14:Free State.mkveed6:lengthi95534e4:pathl4:Subs19:Spanish - Latin.srteed6:lengthi99638e4:pathl4:Subs10:French.srteed6:lengthi99786e4:pathl4:Subs11:English.srteed6:lengthi113036e4:pathl4:Subs16:English(SDH).srteee4:name10:Free State12:piece lengthi262144e6:pieces288880: ......
 .....
:privatei1e6:source29:http://yourprivayetracker.comee

**Comment by drizt** [comment]
https://wiki.theory.org/BitTorrentSpecification#Dictionaries

> Keys must be strings and appear in sorted order (sorted as raw strings, not alphanumerics). The strings should be compared using a binary comparison, not a culture-specific "natural" comparison. 

I see that you give me incorrect format file.

**Comment by cfpp2p** [comment]
> I see that you give me incorrect format file.

Yes, that's the problem. So I suppose that it doesn't make any sense to have _torrent-file-editor_ manipulate invalid torrents without creating valid data in the end. There's a problem stemming from BEP3 20-Oct-2012 update which allows inconsistency

> However, when some 'private trackers' started to generate erroneous .torrents, the chosen resolution was to hash the unsorted info dict as it is. BEP3 has been updated to reflect this.

and thus by doing so allowing torrent clients to use invalid torrents but requiring torrent creators/editors to have

> ...sorted order (sorted as raw strings, not alphanumerics)...

If _torrent-file-editor_ is used (while excluding `info` dictionary)  on such an invalid torrent there results a valid torrent however the hash changes and no longer can clients using the edited torrent connect to peers of the original torrent even though `info` was not user altered in any way. There lies the inconsistency.

One resolution idea I could see is warning when opening torrents with unsorted info dictionary and option (only _if_ info dictionary is not changed in any way) to save torrent with original invalid info dictionary.

Kind of a mess for such a rare but BEP3 allowed situation.

**Comment by drizt** [comment]
Thanks for such good explanation.

**Comment by drizt** [comment]
Can you ask Arvid to explain this in bencode format description? Such thing is not obviously and leads to incorrect implementaion like as my case. There is ambiguity and it should be resolved.

**Comment by cfpp2p** [comment]
> Thanks for such good explanation.

Thanks for such a good torrent-file-editor.

**Comment by cfpp2p** [comment]
@drizt 

> Can you ask Arvid to explain this in bencode format description?...

I made a request for clarification [#55](https://github.com/drizt/torrent-file-editor/issues/47#ref-issue-204150764) to the guys over at [**BitTorrent Inc.**](https://github.com/bittorrent/bittorrent.org/issues) that is receiving attention.

It seems a patent ambiguity i.e. [Postel's law](https://en.wikipedia.org/wiki/Robustness_principle)

> In other words, code that sends commands or data to other machines (or to other programs on the same machine) should conform completely to the specifications, but code that receives input should accept non-conformant input as long as the meaning is clear.
